git/fsck.go

```diff
@@ -0,0 +1,316 @@
+		errs = append(errs, err)
+		return errs
+	}
+	for _, obj := range reachables {
+		if opts.Verbose {
+			fmt.Printf("Checking %v\n", obj)
+		}
+		if _, ok := corrupted[obj]; ok {
+			addErr(fmt.Errorf("%v corrupt or missing", obj))
+			continue
+		}
+		o, _, err := c.HaveObject(obj)
+		if err != nil {
+			addErr(err)
+			continue
+		}
+		if !o {
+			addErr(fmt.Errorf("%v corrupt or missing", obj))
+			continue
+		}
+
+		switch ty := obj.Type(c); ty {
+		case "commit":
+			if err := verifyCommit(c, opts, CommitID(obj)); err != nil {
+				addErr(fmt.Errorf("error in commit %v: %v", obj, err))
+			}
+		case "tree":
+			if err := verifyTree(c, opts, TreeID(obj)); err != nil {
+				addErr(err)
+			}
+		case "blob":
+			// There's not much to verify for a blob, but it's
+			// a known type.
+		default:
+			addErr(fmt.Errorf("Unknown object type %v", ty))
+		}
+	}
+	return errs
+}
+
+// Verifies the HEAD pointer for fsck.
+func verifyHead(c *Client, stderr io.Writer, opts FsckOptions) error {
+	if opts.Verbose {
+		fmt.Fprintln(stderr, "Checking HEAD link")
+	}
+
+	hfile := c.GitDir.File("HEAD")
+	if !hfile.Exists() {
+		return fmt.Errorf("Missing head link")
+	}
+
+	line, err := hfile.ReadFirstLine()
+	if err != nil {
+		// this shouldn't happen since we already verified it exists
+		return err
+	}
+
+	sha1, err := Sha1FromString(line)
+	if err != nil {
+		// we couldn't convert it to a sha1, so it must be a ref
+		// pointer and should point to a head (not a tag or a remote)
+		if !strings.HasPrefix(line, "ref: refs/heads") {
+			return fmt.Errorf("error: HEAD points to something strange")
+		}
+		return nil
+	}
+
+	// We could convert the line to a Sha1, it's a detached head.
+	if sha1 == (Sha1{}) {
+		return fmt.Errorf("error: HEAD: detached HEAD points at nothing")
+	}
+	have, _, err := c.HaveObject(sha1)
+	if err != nil || !have {
+		return fmt.Errorf("error: invalid sha1 pointer %v", sha1)
+	}
+	return nil
+}
+
+// Verifies a commit for fsck or rev-parse --verify-objects
+func verifyCommit(c *Client, opts FsckOptions, cmt CommitID) error {
+	obj, err := c.GetCommitObject(cmt)
+	if err != nil {
+		return err
+	}
+
+	validatePerson := func(typ string) error {
+		s := obj.GetHeader(typ)
+		// 0 = whole match
+		// 1 = name
+		// 2 = email
+		// 3 = timestamp
+		personRe := regexp.MustCompile(`(.*?)\<(.*?)\>(.*)`)
+		pieces := personRe.FindStringSubmatch(s)
+		if len(pieces) != 4 {
+			// This is mostly just to get the same error messages
+			// as git when running the official test suite"
+			// "foo asdf> 1234" is reported as bad name
+			// "foo 1234" is reported as bad email.
+			if strings.Count(s, ">") == 0 {
+				return fmt.Errorf("missingEmail: invalid %v line - missing email", typ)
+			}
+			return fmt.Errorf("badName: invalid %v line - bad name", typ)
+		}
+		if strings.Count(pieces[1], ">") > 0 {
+			return fmt.Errorf("badName: invalid %v line - bad name", typ)
+		}
+		//fmt.Printf(`"%v" "%v" "%v"`+"\n", pieces[1], pieces[2], pieces[3])
```

Comment:
Is it worth logging this?

Reply:
I think it would be too verbose to log.. it was mostly just for figuring out which parts of the regex need to be greedy to pass the fsck tests.